run-database-tests

```diff
@@ -0,0 +1,152 @@
+
+    matcher = re.compile("database_args=([^:]*):([^:]*):([^:]*):([^\n:]*)")
+    def parsedbconfig(config):
+        debug("Parsing %s" % config)
+        with open(config) as fp:
+            for line in fp:
+                m = matcher.match(line)
+                if m != None:
+                    return (m.group(1),m.group(2),m.group(3),m.group(4))
+        return (None,None,None,None)
+
+    drivermatcher = re.compile("database_driver=([^\n]*)")
+    def parsedbdriver(config):
+        debug("Parsing %s" % config)
+        with open(config) as fp:
+            for line in fp:
+                m = drivermatcher.match(line)
+                if m != None:
+                    if m.group(1) == "postgresql":
+                        return ("pgsql")
+                    else:
+                        return (m.group(1))
+        return None
+
+    def prepare(db,dbconfig,setup):
+        (host, port, user, passwd) = dbconfig
+
+        # if a specialized setup file exists for this database, use it
+        sqlfile = "%s.%s" % (setup,db)
+        # otherwise, use the plain sql one
+        if not(os.path.exists("%s.%s" % (setup,db))):
+            sqlfile = "%s.sql" % setup
+
+        debug("setting up %s with %s" % (db,sqlfile))
+        if db == 'pgsql':
+            cmnd = "PGPASSWORD=%s psql -h %s -p %s %s -v ON_ERROR_STOP=1 -q -d links -f %s" % (passwd, host, port, user, sqlfile)
+        elif db == 'mysql':
+            cmnd = "mysql links -h %s -P %s -u %s -p%s  < %s" % (host, port, user,passwd,sqlfile)
+        elif db == 'sqlite3':
+            cmnd = "sqlite3 links < %s" % sqlfile
+        else:
+            raise SystemExit("Unsupported database %s" % db)
+        debug(cmnd)
+        returncode = os.system(cmnd)
+        if returncode != 0:
+            raise SystemExit("Database setup failed: %s for database %s" % (sqlfile,db))
+
+    def run(db,test):
+        debug("running %s %s" % (db,test))
+        dbconfig = "config." + db
+        testfile = test + ".links"
+        returncode = linkscmd(dbconfig,testfile)
+        if returncode != 0:
+            print("!FAILURE: Test %s on database %s" % (test, db))
+            return 1
+        else:
+            print(" SUCCESS: Test %s on database %s" % (test, db))
+            return 0
+
+
+    def runtests(db,params):
+        failures = 0
+        print("%s: Setting up %s" % (working_dir,db))
+        for setup in config['setups']:
+            prepare(db,params,setup)
+        print("%s: Testing %s" % (working_dir,db))
+        for test in config['tests']:
+            if run(db,test) > 0:
+                failures = failures + 1
+        print("%s: %d failures for %s" % (working_dir, failures, db))
+        return failures > 0
+
+    os.chdir(working_dir)
+
+    with open(config_file) as fp:
+        config = json.load(fp)
+
+    debug(config)
+
+    if args.database == "all":
+        dbconfigs = [(db, parsedbconfig("config.%s" % db)) for db in config['databases']]
```

Comment:
How does this work in the case of `(None, None, None, None)` being returned?

Reply:
In the case of sqlite3, this is fine because no configuration is needed, the sqlite3 just looks for a sqlite3 database file in the filename described by the database name.

In that case, if you're trying to run mysql or postgres tests, the configuration is wrong and you get an error like "Expected string, got None".  An error that says the configuration is wrong would be better, btu configurations can also be wrong for other reasons that are more work to check, so  I'm OK with leaving this particular yak a little unshaven.

(It would probably be better to generalize this so that we can point to the location of the sqlite3 database file in the config file, but this isn't currently supported.)